File: src/chain-observers/filters/internal-market-filter.ts
import type { WorkerBee } from "../../bot";
import { OperationClassifier } from "../classifiers";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { DataEvaluationContext } from "../factories/data-evaluation-context";
import { FilterBase } from "./filter-base";

export class InternalMarketFilter extends FilterBase {
  public constructor(
    worker: WorkerBee
  ) {
    super(worker);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      OperationClassifier
    ];
  }

  public async match(data: DataEvaluationContext): Promise<boolean> {
    const { operationsPerType } = await data.get(OperationClassifier);

    for(const {} of (operationsPerType.limit_order_create2_operation ?? []))
      return true;

    for(const {} of (operationsPerType.limit_order_cancel_operation ?? []))
      return true;

    for(const {} of (operationsPerType.limit_order_create_operation ?? []))
      return true;

    return false;
  }
}

File: python/workerbee/chain_observers/filters/internal_market_filter.py
"""InternalMarketFilter — matches limit order operations.

Mirrors src/chain-observers/filters/internal-market-filter.ts.

Uses OperationClassifier. Matches limit_order_create / limit_order_create2 /
limit_order_cancel operations.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .filter_base import FilterBase

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import DataEvaluationContext


class InternalMarketFilter(FilterBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        from ..classifiers.operation_classifier import OperationClassifier

        return [OperationClassifier]

    async def match(self, data: DataEvaluationContext) -> bool:
        from ..classifiers.operation_classifier import OperationClassifier

        result = await data.get(OperationClassifier)
        operations_per_type = result["operations_per_type"]

        return bool(
            operations_per_type.get("limit_order_create2_operation")
            or operations_per_type.get("limit_order_cancel_operation")
            or operations_per_type.get("limit_order_create_operation"),
        )
